discord/ext/menus/__init__.py

```diff
@@ -859,6 +865,151 @@ async def format_page(self, menu, page):
         """
         raise NotImplementedError
 
+class PageDirection(Enum):
+    before = -1
+    after = +1
+
+class PageSpecifier:
+    """A data class that specifies a page for a KeysetPageSource.
+
+    Attributes
+    -----------
+    direction: :class:`PageDirection`:
+        Whether to request the page before or after the last page.
+    reference: Any:
+        The next page should be requested relative to this object, which was returned
+        by :meth:`KeysetPageSource.get_page`.
+        None indicates either the first or the last page.
+    """
+
+    def __init__(self, direction, reference):
+        self.direction = direction
+        self.reference = reference
+
+    def __repr__(self):
+        return '{0.__class__.__qualname__}({0.direction!r}, {0.reference!r})'.format(self)
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, type(self))
+            and self.direction is other.direction
+            and self.reference == other.reference
+        )
+
+    @classmethod
+    def first(cls):
+        return cls(PageDirection.after, None)
+
+    @classmethod
+    def last(cls):
+        return cls(PageDirection.before, None)
+
+class KeysetPageSource:
```

Comment:
I don't understand why this is copy pasted from the actual PageSource interface.